File: lib/models/stock_receipt.dart
class StockReceipt {
  final int stockReceiptId;
  final int productId;
  final int supplierId;
  final int recordedBy;
  final int quantityReceived;
  final int quantityDamaged;
  final DateTime receiptDate;
  final String? notes;
  final String? productName;
  final String? productSku;
  final String? productSerialNo;
  final String? supplierName;
  final String? recordedByUsername;

  StockReceipt({
    required this.stockReceiptId,
    required this.productId,
    required this.supplierId,
    required this.recordedBy,
    required this.quantityReceived,
    required this.quantityDamaged,
    required this.receiptDate,
    this.notes,
    this.productName,
    this.productSku,
    this.productSerialNo,
    this.supplierName,
    this.recordedByUsername,
  });

  factory StockReceipt.fromJson(Map<String, dynamic> json) {
    final productData =
        (json['product'] as Map?)?.cast<String, dynamic>() ?? const {};
    final supplierData =
        (json['supplier'] as Map?)?.cast<String, dynamic>() ?? const {};
    final userData =
        (json['user'] as Map?)?.cast<String, dynamic>() ?? const {};

    int? intValue(dynamic value) {
      if (value is num) {
      return value.toInt();
      }
      return int.tryParse(value?.toString() ?? '');
    }

    final receiptDateValue = json['receipt_date'] ?? json['received_at'];

    return StockReceipt(
      stockReceiptId:
        intValue(json['receipt_id'] ?? json['stock_receipt_id'] ?? json['id']) ??
        0,
      productId: intValue(json['product_id']) ?? 0,
      supplierId: intValue(json['supplier_id']) ?? 0,
      recordedBy: intValue(json['recorded_by']) ?? 0,
      quantityReceived: intValue(json['quantity_received'] ?? json['quantity']) ?? 0,
      quantityDamaged: intValue(json['quantity_damaged']) ?? 0,
      receiptDate: DateTime.parse(receiptDateValue.toString()),
      notes: json['notes']?.toString() ?? json['reference_no']?.toString(),
      productName:
        (productData['product_name'] ?? productData['name'] ?? json['product_name'])
          ?.toString(),
      productSku: (productData['sku'] ?? json['sku'])?.toString(),
      productSerialNo: productData['serial_no']?.toString(),
      supplierName:
          supplierData['supplier_name'] as String? ??
        supplierData['name'] as String? ??
          json['supplier_name'] as String?,
      recordedByUsername:
          userData['username'] as String? ??
          json['recorded_by_username'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'receipt_id': stockReceiptId,
      'product_id': productId,
      'supplier_id': supplierId,
      'recorded_by': recordedBy,
      'quantity_received': quantityReceived,
      'quantity_damaged': quantityDamaged,
      'receipt_date': receiptDate.toIso8601String(),
      'notes': notes,
    };
  }
}

File: app/models/stock_receipt.py
from datetime import datetime
from decimal import Decimal

from sqlalchemy import DateTime, ForeignKey, Integer, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class StockReceipt(Base):
    __tablename__ = "stock_receipt"

    id: Mapped[int] = mapped_column("receipt_id", Integer, primary_key=True, index=True)
    product_id: Mapped[int] = mapped_column(Integer, ForeignKey("product.product_id"), nullable=False, index=True)
    supplier_id: Mapped[int] = mapped_column(Integer, ForeignKey("supplier.supplier_id"), nullable=False, index=True)
    recorded_by: Mapped[int] = mapped_column(Integer, ForeignKey("user.user_id", ondelete="CASCADE"), nullable=False, index=True)
    quantity: Mapped[int] = mapped_column("quantity_received", Integer, nullable=False)
    quantity_damaged: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    received_at: Mapped[datetime] = mapped_column("receipt_date", DateTime(timezone=False), server_default=func.now(), nullable=False)
    reference_no: Mapped[str | None] = mapped_column("notes", Text, nullable=True)

    product: Mapped["Product"] = relationship("Product", back_populates="stock_receipts")
    supplier: Mapped["Supplier"] = relationship("Supplier", back_populates="stock_receipts")
    user: Mapped["User"] = relationship("User")

    @property
    def unit_cost(self) -> Decimal:
        return Decimal("0.00")

    @property
    def recorded_by_username(self) -> str | None:
        return self.user.full_name if self.user else None
